### PR TITLE
Added flood fill in Coconut

### PR DESCRIPTION
**Pull request by Amaras**
Here is the new Coconut code for the flood fill chapter.
I don't see how to make the code more functional, sadly, as it's a really iterative method.

There might be a merge conflict with the C code for now, but in the meantime, this PR is waiting for review.

### CODE REVIEW

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    if new_value == old_value:
+        return
+    colour(canvas, location, old_value, new_value)
+
+    for neighbour in find_neighbours(canvas, location, old_value, new_value):
+        recursive_fill(canvas, neighbour, old_value, new_value)
+
+
+if __name__ == '__main__':
+    # Testing setup
```
Comment (Trashtalk217) [inline contents/flood_fill/code/coconut/flood_fill.coco:74]: Maybe try to make the tests similair to the Julia tests.

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    grid[:2,] = 0
+
+    stack_fill(grid, starting_location, 0, 1)
+    try:
+        assert (grid == solution_grid).all()
+    except AssertionError:
+        print('F', end='')
+        failure_count += 1
+    else:
+        print('.', end='')
```
Comment (Trashtalk217) [inline contents/flood_fill/code/coconut/flood_fill.coco:109]: The `.` is a bit vague, maybe print something slightly more obvious `S` for succes, or maybe just full words: Success and Failure.

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        return
+
+    if canvas[location] != old_value:
+        return
+    else:
+        canvas[location] = new_value
+
+
+def find_neighbours(canvas, location is Point, old_value, new_value):
+    possible_neighbours = (Point(0, 1), Point(1, 0), Point(0, -1), Point(-1, 0)) |> map$(location.__add__)
```
Comment (Trashtalk217) [inline contents/flood_fill/code/coconut/flood_fill.coco:26]: PEP8 recommends a maximum line length of 79. And while I think 79 is a very low for line length, this line in particular is a bit long.

**Review by Trashtalk217 — CHANGES_REQUESTED** [review]
The .md file still needs some work, but the code compiles runs and succeeds.
It also feels like the Coconut features can be used slightly more.

In conclusion: I've learned a new dialect today and it's pretty neat!

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    def __add__(self, other is Point) = Point(self.x + other.x, self.y + other.y)
+
+
+# This function is necessary, because negative indices wrap around the
+# array in Coconut.
+def inbounds(canvas_shape, location is Point) =
+    min(location) >= 0 and location.x < canvas_shape[0] and location.y < canvas_shape[1]
+
+
+def colour(canvas, location is Point, old_value, new_value):
```
[the remark this answers is not in the document]
Comment (Amaras) [reply contents/flood_fill/code/coconut/flood_fill.coco:15]: I need to rework the code to remove that function, indeed

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+def stack_fill(canvas, location is Point, old_value, new_value):
+    if new_value == old_value:
+        return
+
+    stack = [location]
+
+    while stack:
+        current_location = stack.pop()
+        colour(canvas, current_location, old_value, new_value)
+        for neighbour in find_neighbours(canvas, current_location, old_value,
```
[the remark this answers is not in the document]
Comment (Amaras) [reply contents/flood_fill/code/coconut/flood_fill.coco:42]: For this one, I think it's a good idea

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        return
+
+    queue = deque()
+    queue.append(location)
+
+    colour(canvas, location, old_value, new_value)
+
+    while queue:
+        current_location = queue.popleft()
+        for neighbour in find_neighbours(canvas, current_location, old_value,
```
[the remark this answers is not in the document]
Comment (Amaras) [reply contents/flood_fill/code/coconut/flood_fill.coco:58]: I am not sure I see how to do that right away. I probably need to work on it a bit more, but that seems out of my capabilities for now.

File: contents/flood_fill/code/coconut/flood_fill.coco (hunk 1 above)
Comment (Amaras) [reply contents/flood_fill/code/coconut/flood_fill.coco:74]: What do you mean "similar to the Julia tests"?

File: contents/flood_fill/code/coconut/flood_fill.coco (hunk 2 above)
Comment (Amaras) [reply contents/flood_fill/code/coconut/flood_fill.coco:109]: I made it to look like how `unittest` runs tests, so that's a revertible design choice

**Review by Amaras — COMMENTED** [review]
So, I know it needs rework, but some points you raised are, I think, not really valid.
I'm glad to see more people work with Coconut, though :D

File: contents/flood_fill/code/coconut/flood_fill.coco [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            queue.append(neighbour)
+            colour(canvas, neighbour, old_value, new_value)
+
+
+def recursive_fill(canvas, location is Point, old_value, new_value):
+    if new_value == old_value:
+        return
+    colour(canvas, location, old_value, new_value)
+
+    for neighbour in find_neighbours(canvas, location, old_value, new_value):
```
[the remark this answers is not in the document]
Comment (Trashtalk217) [reply contents/flood_fill/code/coconut/flood_fill.coco:69]: It's definitely a little too long, splitting it up in `neighbours = find_neighbours(...)` and `neighbours |> map$(...)` is probably a good idea.

About the confusion bit. I think that maps are as intuitive as for-loops. It's just that a lot of people only know for-loops, so maps seem weird to them. I may be slightly (very) biased here, as I have some experience with functional programming, so keep that in mind.

File: contents/flood_fill/code/coconut/flood_fill.coco (hunk 1 above)
Comment (Trashtalk217) [reply contents/flood_fill/code/coconut/flood_fill.coco:74]: Disregard the thing I said. I meant to say that you should probably use the same test as in the Julia implementation, namely:
```
grid = zeros(5,5)
grid[3,:] .= 1
```
But I just realize that you are already doing so. It's just that Julia indexes it's arrays starting at 1. Got confused there.

**Comment by Trashtalk217** [comment]
Sometimes I am just wrong, so feel free to call me out. It's only my first time touching the language after all.

**Comment by Amaras** [comment]
Closing to redo because I did something stupid.

**Pull request closed without merging**